# 4-3. Spring Boot - 예외처리

Spring Boot 동작방식

![Web 4-3](https://github.com/devchae97/TIL/blob/master/img/Web%204-3.png?raw=true)

Reqest가 들어온다면, Filter를 거쳐 DispacherServlet 도착

Handler Mapping을 통해 어떠한 주소를 가진 컨트롤러에 매핑 해야하는지 결정 후

Handler Interceptor를 통과해 Controller에 해당 Request가 들어오면 Service 및 DB 처리 후 응답

허나, Controller, 비즈니스 로직을 처리하는 구간에서 예외가 발생하게 되면 Exception Handler가 해당 예외를 캐치해 에러에 대한 응답

<br/>

- 예외를 처리할 경우는 두가지로 분류

  1. 예상하지 못했던 예외가 발생하는 경우 : Exception Handler를 통해서 응답

  2. 예측했던 예외를 일부러 터트리는 경우 : (추후 다시 다루기)

<br/>

> Spirng initializr
>
> Gradle - Groovy / Java / 2.7.8 / exception / Jar / 11 / + Lombok, Spring Web

예외발생

```java
package com.example.exception.Controller;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api")
public class RestApiController {

    @GetMapping(path = "")
    public void hello(){
        throw new RuntimeException("run time exception call");
    }
}
```

:arrow_right: 500 error 발생

<br/>

예외발생 : IndexOutOfBoundsException (가장 흔하게 접하는 예외 중 한가지)

```java
package com.example.exception.Controller;

import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@Slf4j
@RestController
@RequestMapping("/api")
public class RestApiController {

    @GetMapping(path = "")
    public void hello(){
        var list = List.of("hello");
        var element = list.get(1);
        log.info("element : {}", element);
    }
}
// ERROR 197692 --- [nio-8080-exec-1] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is java.lang.IndexOutOfBoundsException: Index: 1 Size: 1] with root cause
// java.lang.IndexOutOfBoundsException: Index: 1 Size: 1
```

:arrow_right: 500 error 발생, Spring Boot Web에서는 자동적으로 예외가 발생할 경우 "Internal Server Error" 메세지 응답

<br/>

이러한 예외 부분을 하나하나 try catch 문으로 묶어주는 방법은 비효율적이며 찾기 또한 어려움 > ExceptionHandler 사용

```java
package com.example.exception.Exception;

import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

@Slf4j
@RestControllerAdvice // 해당 클래스는 RestAPI가 사용하는 곳에 예외 발생을 감지
public class RestApiExceptionHandler {

    @ExceptionHandler(value = {Exception.class}) // exception 클래스는 전범위
    public ResponseEntity exception(Exception e){
        log.error("RestApiExceptionHanlder", e);
        return ResponseEntity.status(200).build();
    }
}
// ERROR 130336 --- [nio-8080-exec-1] c.e.e.Exception.RestApiExceptionHandler  : RestApiExceptionHanlder
// java.lang.IndexOutOfBoundsException: Index: 1 Size: 1
```

:arrow_right: 200 정상응답, 해당 Exception을 찍은 곳이 ExceptionHandler의 exception 메서드

<br/>

Exception 메서드 특정시키기

```java
package com.example.exception.Exception;

import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

@Slf4j
@RestControllerAdvice // 해당 클래스는 RestAPI가 사용하는 곳에 예외 발생을 감지
public class RestApiExceptionHandler {

    @ExceptionHandler(value = {Exception.class})
    public ResponseEntity exception(Exception e){
        log.error("RestApiExceptionHanlder", e);
        return ResponseEntity.status(200).build();
    }

    @ExceptionHandler(value = {IndexOutOfBoundsException.class})
    public ResponseEntity outOfBound(IndexOutOfBoundsException e){
        log.error("IndexOutOfBoundsException", e);
        return ResponseEntity.status(200).build();
    }
}

// ERROR 211596 --- [nio-8080-exec-1] c.e.e.Exception.RestApiExceptionHandler  : IndexOutOfBoundsException
// java.lang.IndexOutOfBoundsException: Index: 1 Size: 1
```

:arrow_right: 200 정상응답, 명확하게 IndexOutOfBoundException은 아래에서 잡겠다고 지정했기에 outOfBound 메서드

<br/>



> Reference
>
> Fastcampus : Signature Backend